client/client-multi/client-cpp/impl/channel/KaaChannelManager.cpp

```diff
@@ -67,63 +68,112 @@ void KaaChannelManager::setFailoverStrategy(IFailoverStrategyPtr strategy) {
+                                                        % LoggingUtils::toString(reason)
+                                                        % LoggingUtils::toString(*connectionInfo));
+        return;
     }
 
+    KAA_LOG_TRACE(boost::format("Processing '%s' failover for %s")
+                                                    % LoggingUtils::toString(reason)
+                                                    % LoggingUtils::toString(*connectionInfo));
+
     if (connectionInfo->isFailedState()) {
-        KAA_LOG_DEBUG("Connection already failed. Ignoring connection failover!");
+        KAA_LOG_TRACE(boost::format("Ignoring failover: marked as already failed %s")
+                                                % LoggingUtils::toString(*connectionInfo));
         return;
     } else {
         connectionInfo->setFailedState();
     }
 
+    checkAuthenticationFailover(reason);
+
     if (connectionInfo->getServerType() == ServerType::BOOTSTRAP) {
-        ITransportConnectionInfoPtr nextConnectionInfo = getNextBootstrapServer(connectionInfo->getTransportId(), false);
-        if (nextConnectionInfo) {
-            onTransportConnectionInfoUpdated(nextConnectionInfo);
-        } else {
-            FailoverStrategyDecision decision = failoverStrategy_->onFailover(reason);
-            switch (decision.getAction()) {
-                 case FailoverStrategyAction::NOOP:
-                     KAA_LOG_WARN("No operation is performed according to failover strategy decision.");
-                     break;
-                 case FailoverStrategyAction::RETRY:
-                 {
-                     std::size_t period = decision.getRetryPeriod();
-                     KAA_LOG_WARN(boost::format("Attempt to reconnect to first bootstrap server will be made in %1% secs "
-                             "according to failover strategy decision.") % period);
-                     bsTransportId_ = connectionInfo->getTransportId();
-                     retryTimer_.stop();
-                     retryTimer_.start(period, [&]
-                         {
-                             onTransportConnectionInfoUpdated(getNextBootstrapServer(bsTransportId_, true));
-                         });
-                     break;
-                 }
-                 case FailoverStrategyAction::STOP_CLIENT:
-                     KAA_LOG_WARN("Stopping client according to failover strategy decision!");
-                     client_->stop();
-                     break;
-                 default:
-                    break;
-             }
-        }
+        onBootstrapServerFailed(connectionInfo, reason);
     } else {
-        bootstrapManager_.useNextOperationsServer(connectionInfo->getTransportId(), reason);
+        bootstrapManager_.onOperationsServerFailed(connectionInfo->getTransportId(), reason);
     }
 }
 
+/*
+ * NOTE: Expects a valid connection info, that is not null and has the type Bootstrap.
```

Comment:
Also, put some assertions, or exception throw in case if types are mismatched. Ideally, use correct parameter type.